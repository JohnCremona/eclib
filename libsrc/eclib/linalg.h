// linalg.h: typedefs for scalar, vector, matrix types
//
// Include this to use any of the vec/mat/svec/smat/subspace/ssubspace classes
//
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

// #define XSTR(x) STR(x)
// #define STR(x) #x
// #ifdef SCALR_OPTION
// #pragma message "The value of SCALAR_OPTION (before): " XSTR(SCALAR_OPTION)
// #else
// #pragma message "SCALAR_OPTION (before) is undefined"
// #endif

#include "modulus.h"
#include "vector.h"
#include "svector.h"
#include "matrix.h"
#include "smatrix.h"
#include "subspace.h"
#include "ssubspace.h"
#include "smatrix_elim.h"

// SCALAR_OPTION may be set to 1 (int), 2 (long), or 3 (bigint) by the
// user.  This determines whether vec, mat, ... default to vec_i,
// mat_i, ... or vac_l, mat_l, ..., or vec_m, mat_m, ... in user code.
// The default value set here is what these abbreviations mean when
// the user does not specify a different option.  NB The default also
// controls the option with which the library is built, which affects
// the linear algebra scalar type used in the modular symbol code
// (which is not currently templated).

#ifndef SCALAR_OPTION
//#define SCALAR_OPTION 1  // int
//#define SCALAR_OPTION 2  // long
#define SCALAR_OPTION 3  // bigint
#endif

// #pragma message "The value of SCALAR_OPTION (after): " XSTR(SCALAR_OPTION)

#undef scalar_type // since this file may be included more than once

#define PRIME30  1073741789  // = largest p such that p < 2^30.

#if (SCALAR_OPTION==2) // long
// #pragma message "In branch with SCALAR_OPTION 2"
#define scalar_type string("long")
typedef long scalar;
typedef vec_l vec;
typedef mat_l mat;
typedef subspace_l subspace;
typedef ssubspace_l ssubspace;
typedef svec_l svec;
typedef smat_l smat;
typedef smat_l_elim smat_elim;
typedef form_finder_l form_finder;
#define to_vec to_vec_l
#define to_mat to_mat_l
#else
#if (SCALAR_OPTION==3) // bigint
// #pragma message "In branch with SCALAR_OPTION 3"
#define scalar_type string("bigint")
typedef bigint scalar;
typedef vec_m vec;
typedef mat_m mat;
typedef subspace_m subspace;
typedef ssubspace_m ssubspace;
typedef svec_m svec;
typedef smat_m smat;
typedef smat_m_elim smat_elim;
typedef form_finder_m form_finder;
#define to_vec to_vec_m
#define to_mat to_mat_m
#else
#if (SCALAR_OPTION==1) // int
// #pragma message "In branch with SCALAR_OPTION 1"
#define scalar_type string("int")
typedef int scalar;
typedef vec_i vec;
typedef mat_i mat;
typedef subspace_i subspace;
typedef ssubspace_i ssubspace;
typedef svec_i svec;
typedef smat_i smat;
typedef smat_i_elim smat_elim;
typedef form_finder_i form_finder;
#define to_vec to_vec_i
#define to_mat to_mat_i
#else
#pragma message "In branch with SCALAR_OPTION not 1, 2, or 3"
#define scalar_type string("undefined")
#endif
#endif
#endif
