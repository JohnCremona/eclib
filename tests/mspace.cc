// mspace.cc: test program for subspace class with bigint scalars
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include <eclib/linalg.h>

const bigint modulus(default_modulus<bigint>());

int main()
{
cin.flags( cin.flags() | ios::dec );

cout << "\nM-Subspace package test program\n\n";

int i,r=0;

while (cout << "Enter size of square matrix M: ", cin >> r, r>0 )
{
  mat_m m(r,r);
  cout << "Enter entries of M: ";
  cin >> m;
  cout << " M = \n" << m << endl;
  cout << "Trace(M) = " << m.trace() << endl;
  //
  mat_m mpower=m;
  for (i=2; i<=r; i++)
    {mpower=mpower*m;
     cout << "m^" << i << " = \n" << mpower << endl;
     cout << "Trace(m^" << i << ") = " << mpower.trace() << endl;
   }
//
  {
    vector<bigint> cp = m.charpoly();
    cout << "char. poly. of m has coefficients " << cp << endl;
  }
  cout << "det(M) = " << m.determinant() << endl;
  cout << "rank(M) = " << m.rank() << endl;
  cout << "nullity(M) = " << m.nullity() << endl;
  {
    subspace_m ker = kernel(m);
    mat_m kerbasis = basis(ker);
    cout << "kernel(m) has basis\n" << kerbasis << endl;
    vec_i kerpivs = pivots(ker);
    cout << "pivots: " << kerpivs << "\n";
    bigint kerdenom = denom(ker);
    cout << "denom:  " << kerdenom  << "\n";
  }
  {
    subspace_m im = image(m);
    cout << "image(m) has basis\n" << basis(im) << endl;
    cout << "pivots: " << pivots(im) << "\n";
    cout << "denom:  " << denom(im)  << "\n";
  }
  {
    bigint lambda;
    cout << "Enter lambda: "; cin >> lambda;
    subspace_m elambda = eigenspace(m,lambda);
    cout << "eigenspace for lambda = " << lambda << " has basis\n" << basis(elambda) << endl;
    cout << "with dimension " << dim(elambda) << endl;
    cout << "\nNow repeating eigenspace calculation modulo " << modulus << endl;
    subspace_m elp;
    lift(peigenspace(m,lambda,modulus),modulus, elp);
    cout << "eigenspace for lambda has basis\n" << basis(elp) << endl;
    cout << "with dimension " << dim(elp) << endl;
  }
}
 cout<<endl;
}
