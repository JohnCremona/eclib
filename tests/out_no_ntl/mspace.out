
M-Subspace package test program

Enter size of square matrix M: Enter entries of M:  M = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Trace(M) = 15
m^2 = 
[[30,36,42],
[66,81,96],
[102,126,150]]
Trace(m^2) = 261
m^3 = 
[[468,576,684],
[1062,1305,1548],
[1656,2034,2412]]
Trace(m^3) = 4185
char. poly. of m has coefficients [ 0 -18 -15 1 ]
det(M) = 0
rank(M) = 2
nullity(M) = 1
kernel(m) has basis

[[1],
[-2],
[1]]
pivots: [3]
denom:  1
image(m) has basis

[[1,0],
[0,1],
[-1,2]]
pivots: [1,2]
denom:  1
Enter lambda: eigenspace for lambda = 1 has basis

[[],
[],
[]]
with dimension 0

Now repeating eigenspace calculation modulo 6074000003
eigenspace for lambda has basis

[[],
[],
[]]
with dimension 0
Enter size of square matrix M: Enter entries of M:  M = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Trace(M) = 15
m^2 = 
[[30,36,42],
[66,81,96],
[102,126,150]]
Trace(m^2) = 261
m^3 = 
[[468,576,684],
[1062,1305,1548],
[1656,2034,2412]]
Trace(m^3) = 4185
char. poly. of m has coefficients [ 0 -18 -15 1 ]
det(M) = 0
rank(M) = 2
nullity(M) = 1
kernel(m) has basis

[[1],
[-2],
[1]]
pivots: [3]
denom:  1
image(m) has basis

[[1,0],
[0,1],
[-1,2]]
pivots: [1,2]
denom:  1
Enter lambda: eigenspace for lambda = 0 has basis

[[1],
[-2],
[1]]
with dimension 1

Now repeating eigenspace calculation modulo 6074000003
eigenspace for lambda has basis

[[1],
[-2],
[1]]
with dimension 1
Enter size of square matrix M: 
